### PR TITLE
Resubmit optimization uniq to count

### PR DESCRIPTION
**Pull request by JackyWoo**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
This PR will fix UBsan test error [here](https://github.com/ClickHouse/ClickHouse/pull/54566)

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 70a262a775f2ba5ea7eabf4a199f66ef66f607b3 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/54568/70a262a775f2ba5ea7eabf4a199f66ef66f607b3/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>✅ success</td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>❌ failure</td></tr>
<tbody>
</table>

**Comment by JackyWoo** [comment]
@alexey-milovidov Please review the PR when you are free.

**Comment by JackyWoo** [comment]
@alexey-milovidov The last commit will fix the UBsan test error https://github.com/ClickHouse/ClickHouse/pull/54566. 

There are some errors in CI, but maybe not caused by the PR.

File: src/Analyzer/Passes/UniqToCountPass.cpp [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+namespace DB
+{
+
+namespace
+{
+
+bool matchFnUniq(String func_name)
+{
+    auto name = Poco::toLower(func_name);
+    return name == "uniq" || name == "uniqHLL12" || name == "uniqExact" || name == "uniqTheta" || name == "uniqCombined"
```
Comment (alexey-milovidov) [inline src/Analyzer/Passes/UniqToCountPass.cpp:20]: After we made function lowercase, there is no point comparing it to `uniqHLL12`, `uniqExact`, `uniqTheta`, `uniqCombined`

File: src/Interpreters/RewriteUniqToCountVisitor.cpp [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+using Aliases = std::unordered_map<String, ASTPtr>;
+
+namespace
+{
+
+bool matchFnUniq(String func_name)
+{
+    auto name = Poco::toLower(func_name);
+    return name == "uniq" || name == "uniqHLL12" || name == "uniqExact" || name == "uniqTheta" || name == "uniqCombined"
```
Comment (alexey-milovidov) [inline src/Interpreters/RewriteUniqToCountVisitor.cpp:25]: ?

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
.

File: src/Interpreters/RewriteUniqToCountVisitor.cpp [hunk 3]
```diff
@@ -22,8 +22,8 @@ namespace
 bool matchFnUniq(String func_name)
 {
     auto name = Poco::toLower(func_name);
-    return name == "uniq" || name == "uniqHLL12" || name == "uniqExact" || name == "uniqTheta" || name == "uniqCombined"
-        || name == "uniqCombined64";
+    return name == "uniq" || name == "uniqhll12" || name == "uniqexact" || name == "uniqtheta" || name == "uniqcombined"
```
Comment (alexey-milovidov) [inline src/Interpreters/RewriteUniqToCountVisitor.cpp:25]: I figured out that this is theoretically wrong.

The function `uniqCombined` is not case-insensitive.
A user could register a user-defined function with `uniqcombined` name, and the optimization will be applied incorrectly.

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**